File: sdk/typescript/src/namespaces/skills.ts
/**
 * Skills Namespace API
 *
 * Provides skill management and invocation:
 * - List available skills
 * - Get skill details and metrics
 * - Invoke skills with input data
 */

/**
 * Skill capability types.
 */
export type SkillCapability =
  | 'READ_LOCAL'
  | 'WRITE_LOCAL'
  | 'READ_DATABASE'
  | 'WRITE_DATABASE'
  | 'EXTERNAL_API'
  | 'WEB_SEARCH'
  | 'WEB_FETCH'
  | 'CODE_EXECUTION'
  | 'SHELL_EXECUTION'
  | 'LLM_INFERENCE'
  | 'EMBEDDING'
  | 'DEBATE_CONTEXT'
  | 'EVIDENCE_COLLECTION'
  | 'KNOWLEDGE_QUERY'
  | 'SYSTEM_INFO'
  | 'NETWORK';

/**
 * Skill execution status.
 */
export type SkillStatus =
  | 'SUCCESS'
  | 'FAILURE'
  | 'PARTIAL'
  | 'TIMEOUT'
  | 'RATE_LIMITED'
  | 'PERMISSION_DENIED'
  | 'INVALID_INPUT'
  | 'NOT_IMPLEMENTED';

/**
 * Skill manifest.
 */
export interface SkillManifest {
  name: string;
  version: string;
  description?: string;
  capabilities: SkillCapability[];
  input_schema: Record<string, unknown>;
  output_schema?: Record<string, unknown>;
  tags?: string[];
  rate_limit_per_minute?: number;
  timeout_seconds?: number;
}

/**
 * Skill details (extended manifest).
 */
export interface SkillDetails extends SkillManifest {
  author?: string;
  required_permissions?: string[];
  required_env_vars?: string[];
  max_execution_time_seconds?: number;
  debate_compatible?: boolean;
  requires_debate_context?: boolean;
}

/**
 * Skill metrics.
 */
export interface SkillMetrics {
  skill: string;
  total_invocations: number;
  successful_invocations: number;
  failed_invocations: number;
  average_latency_ms: number;
  last_invoked?: string;
}

/**
 * Invoke skill request.
 */
export interface InvokeSkillRequest {
  /** Skill name to invoke */
  skill: string;
  /** Input data for the skill */
  input?: Record<string, unknown>;
  /** User ID for RBAC */
  user_id?: string;
  /** User permissions */
  permissions?: string[];
  /** Additional metadata/config */
  metadata?: Record<string, unknown>;
  /** Timeout in seconds (default: 30, max: 60) */
  timeout?: number;
}

/**
 * Invoke skill result.
 */
export interface InvokeSkillResult {
  status: 'success' | 'error';
  output?: Record<string, unknown>;
  error?: string;
  execution_time_ms?: number;
  metadata?: Record<string, unknown>;
}

/**
 * Client interface for skills operations.
 */
interface SkillsClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Skills API namespace.
 *
 * Provides methods for managing and invoking skills:
 * - List all available skills
 * - Get skill details and metrics
 * - Invoke skills with custom input
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // List all skills
 * const { skills } = await client.skills.list();
 *
 * // Get skill details
 * const skill = await client.skills.get('web-search');
 *
 * // Invoke a skill
 * const result = await client.skills.invoke({
 *   skill: 'web-search',
 *   input: { query: 'TypeScript best practices' },
 * });
 * ```
 */
export class SkillsAPI {
  constructor(private client: SkillsClientInterface) {}

  /**
   * List all available skills.
   */
  async list(): Promise<{ skills: SkillManifest[]; total: number }> {
    return this.client.request('GET', '/api/skills');
  }

  /**
   * Get skill details.
   */
  async get(name: string): Promise<SkillDetails> {
    return this.client.request('GET', `/api/skills/${encodeURIComponent(name)}`);
  }

  /**
   * Get skill metrics.
   */
  async getMetrics(name: string): Promise<SkillMetrics> {
    return this.client.request('GET', `/api/skills/${encodeURIComponent(name)}/metrics`);
  }

  /**
   * Invoke a skill.
   */
  async invoke(request: InvokeSkillRequest): Promise<InvokeSkillResult> {
    return this.client.request('POST', '/api/skills/invoke', {
      json: request as unknown as Record<string, unknown>,
    });
  }
}

File: sdk/python/aragora/namespaces/skills.py
"""
Skills Namespace API

Provides skill management and invocation:
- List available skills
- Get skill details and metrics
- Invoke skills with input data

Features:
- Skill discovery and inspection
- Skill execution with RBAC
- Metrics and monitoring
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


SkillCapability = Literal[
    "READ_LOCAL",
    "WRITE_LOCAL",
    "READ_DATABASE",
    "WRITE_DATABASE",
    "EXTERNAL_API",
    "WEB_SEARCH",
    "WEB_FETCH",
    "CODE_EXECUTION",
    "SHELL_EXECUTION",
    "LLM_INFERENCE",
    "EMBEDDING",
    "DEBATE_CONTEXT",
    "EVIDENCE_COLLECTION",
    "KNOWLEDGE_QUERY",
    "SYSTEM_INFO",
    "NETWORK",
]

SkillStatus = Literal[
    "SUCCESS",
    "FAILURE",
    "PARTIAL",
    "TIMEOUT",
    "RATE_LIMITED",
    "PERMISSION_DENIED",
    "INVALID_INPUT",
    "NOT_IMPLEMENTED",
]


class SkillsAPI:
    """
    Synchronous Skills API.

    Provides methods for managing and invoking skills:
    - List all available skills
    - Get skill details and metrics
    - Invoke skills with custom input

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> skills = client.skills.list()
        >>> result = client.skills.invoke(skill="web-search", input={"query": "TypeScript"})
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(self) -> dict[str, Any]:
        """
        List all available skills.

        Returns:
            Dict with:
            - skills: List of skill manifests
            - total: Total count
        """
        return self._client.request("GET", "/api/skills")

    def get(self, name: str) -> dict[str, Any]:
        """
        Get skill details.

        Args:
            name: Skill name

        Returns:
            Dict with full skill details including:
            - name, version, description
            - capabilities
            - input_schema, output_schema
            - rate_limit_per_minute
            - timeout_seconds
        """
        return self._client.request("GET", f"/api/skills/{name}")

    def get_metrics(self, name: str) -> dict[str, Any]:
        """
        Get skill metrics.

        Args:
            name: Skill name

        Returns:
            Dict with:
            - skill: Skill name
            - total_invocations: Total calls
            - successful_invocations: Success count
            - failed_invocations: Failure count
            - average_latency_ms: Average latency
            - last_invoked: Last invocation time
        """
        return self._client.request("GET", f"/api/skills/{name}/metrics")

    def invoke(
        self,
        skill: str,
        input: dict[str, Any] | None = None,
        user_id: str | None = None,
        permissions: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        timeout: int | None = None,
    ) -> dict[str, Any]:
        """
        Invoke a skill.

        Args:
            skill: Skill name to invoke
            input: Input data for the skill
            user_id: User ID for RBAC
            permissions: User permissions
            metadata: Additional metadata
            timeout: Timeout in seconds (max 60)

        Returns:
            Dict with:
            - status: success or error
            - output: Skill output
            - error: Error message if failed
            - execution_time_ms: Execution time
        """
        data: dict[str, Any] = {"skill": skill}
        if input:
            data["input"] = input
        if user_id:
            data["user_id"] = user_id
        if permissions:
            data["permissions"] = permissions
        if metadata:
            data["metadata"] = metadata
        if timeout:
            data["timeout"] = timeout
        return self._client.request("POST", "/api/skills/invoke", json=data)


class AsyncSkillsAPI:
    """
    Asynchronous Skills API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     skills = await client.skills.list()
        ...     result = await client.skills.invoke(skill="web-search", input={"query": "AI"})
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(self) -> dict[str, Any]:
        """List all available skills."""
        return await self._client.request("GET", "/api/skills")

    async def get(self, name: str) -> dict[str, Any]:
        """Get skill details."""
        return await self._client.request("GET", f"/api/skills/{name}")

    async def get_metrics(self, name: str) -> dict[str, Any]:
        """Get skill metrics."""
        return await self._client.request("GET", f"/api/skills/{name}/metrics")

    async def invoke(
        self,
        skill: str,
        input: dict[str, Any] | None = None,
        user_id: str | None = None,
        permissions: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        timeout: int | None = None,
    ) -> dict[str, Any]:
        """Invoke a skill."""
        data: dict[str, Any] = {"skill": skill}
        if input:
            data["input"] = input
        if user_id:
            data["user_id"] = user_id
        if permissions:
            data["permissions"] = permissions
        if metadata:
            data["metadata"] = metadata
        if timeout:
            data["timeout"] = timeout
        return await self._client.request("POST", "/api/skills/invoke", json=data)
